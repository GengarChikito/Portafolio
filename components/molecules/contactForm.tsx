import { Form, Input, Button, message } from "antd";

/**
 * Defino las 'props' para este componente.
 * En este caso, no necesito ninguna prop, pero mantengo la interfaz
 * por si decido añadirle algo en el futuro (como un título).
 */
export interface ContactFormProps {}

/**
 * Este es mi componente 'ContactForm', una "molécula".
 * Es un formulario simple de Ant Design para que los visitantes
 * me puedan enviar un mensaje.
 */
export function ContactForm(_: ContactFormProps) {
    // Uso el hook 'useForm' de Ant Design.
    // Esto me da una instancia de 'form' que puedo usar para
    // controlar el formulario, como por ejemplo, para limpiarlo.
    const [form] = Form.useForm();

    /**
     * Esta es la función que se ejecuta cuando el formulario se envía
     * y *pasa* todas las validaciones (es decir, todos los campos
     * requeridos están llenos).
     */
    async function onFinish(): Promise<void> {
        // Por ahora, no estoy enviando el formulario a un backend (es un TODO).
        // Simplemente muestro un mensaje de éxito con 'message.success'.
        message.success("Gracias por tu mensaje. Te responderé pronto.");

        // Uso la instancia 'form' para limpiar (resetear) los campos
        // después de que se envía exitosamente.
        form.resetFields();
    }

    return (
       /* // Aquí empieza el formulario.
        // layout="vertical" pone las etiquetas (label) encima de los campos.
        // Le paso mi instancia 'form' y mi función 'onFinish'.*/
        <Form layout="vertical" form={form} onFinish={onFinish}>

            {/* Campo 1: Nombre */}
            <Form.Item
                label="Nombre"
                name="name"
               /* // 'rules' define las reglas de validación.
                // Aquí, es un campo obligatorio.*/
                rules={[{ required: true, message: "Ingresa tu nombre" }]}
            >
                <Input placeholder="Tu nombre" />
            </Form.Item>

            {/* Campo 2: Correo */}
            <Form.Item
                label="Correo"
                name="email"
                // Añado dos reglas: es obligatorio y debe ser de tipo 'email'.
                rules={[{ required: true, type: "email", message: "Correo válido" }]}
            >
                <Input placeholder="tu@correo.com" />
            </Form.Item>

            {/* Campo 3: Mensaje */}
            <Form.Item
                label="Mensaje"
                name="msg"
                rules={[{ required: true, message: "Escribe tu mensaje" }]}
            >
                {/* Uso 'Input. TextArea' para un campo de texto más grande. */}
                <Input.TextArea rows={5} placeholder="¿En qué puedo ayudarte?" />
            </Form.Item>

            {/* Botón de envío */}
            <Button type="primary" htmlType="submit">
                Enviar
            </Button>
        </Form>
    );
}