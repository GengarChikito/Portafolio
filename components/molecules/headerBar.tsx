import { useState } from "react";
import { Button, Space, Switch, Drawer, Grid, Typography } from "antd";
import {
    MenuOutlined,
    GithubOutlined,
    LinkedinFilled,
    FileTextOutlined,
} from "@ant-design/icons";
import { styles } from "app/styles";

/**
 * Defino las 'props' que este Header va a recibir desde el 'HomeLayout' (el padre).
 * - 'dark': El estado booleano actual (¬øest√° oscuro?).
 * - 'onToggleDark': La funci√≥n que el 'HomeLayout' me pasa para *cambiar* ese estado.
 */
export interface HeaderBarProps {
    dark: boolean;
    onToggleDark: (val: boolean) => void;
}

/**
 * Este es mi componente 'HeaderBar', una "mol√©cula" clave.
 * Es la barra de navegaci√≥n superior que se mantiene fija ('sticky').
 * Maneja su propio estado para el men√∫ m√≥vil (el 'Drawer')
 * y renderiza condicionalmente la navegaci√≥n o un bot√≥n de men√∫
 * dependiendo del tama√±o de la pantalla.
 */
export function HeaderBar({ dark, onToggleDark }: HeaderBarProps) {
    // ---- ESTADO ----

    // Este es el estado para el men√∫ 'Drawer' (el panel lateral) en m√≥vil.
    // 'open' ser√° 'true' si el men√∫ est√° abierto.
    const [open, setOpen] = useState(false);

    // ---- L√ìGICA RESPONSIVE ----

    // Uso el hook 'useBreakpoint' de Ant Design.
    // 'screens' es un objeto (ej. { md: true, sm: false }) que me dice
    // qu√© "breakpoints" (puntos de quiebre) est√°n activos.
    const screens = Grid.useBreakpoint();

    // Defino 'isMobile' como "true" si la pantalla es m√°s peque√±a que 'md' (medium).
    // Si 'screens.md' es 'true', significa que estamos en escritorio.
    // Por lo tanto, 'isMobile' es 'false'.
    const isMobile = !screens.md;

    // ---- BLOQUES DE JSX ----

    // Defino mis enlaces de navegaci√≥n en una variable.
    // As√≠ puedo reutilizar este bloque de JSX tanto en el header
    // de escritorio como dentro del Drawer m√≥vil.
    const navLinks = (
        <Space size={[16, 12]} wrap>
            <a className="site-nav-link" href="#inicio">Inicio</a>
            <a className="site-nav-link" href="#proyectos">Proyectos</a>
            <a className="site-nav-link" href="#experiencia">Experiencia</a>
            <a className="site-nav-link" href="#educacion">Educaci√≥n</a>
            <a className="site-nav-link" href="#contacto">Contacto</a>
        </Space>
    );

    // Defino mis botones de acci√≥n (Switch, CV, GitHub, LinkedIn)
    // en otra variable para poder reutilizarlos en ambos modos.
    const actionButtons = (
        <Space size={[8, 8]} wrap>
            {/* Este es el √°tomo 'ThemeSwitch', pero hecho con Ant Design directo.
                Le paso las props 'dark' y 'onToggleDark' que recib√≠ del 'HomeLayout'.
             */}
            <Switch
                checked={dark}
                onChange={onToggleDark}
                checkedChildren="üåô"
                unCheckedChildren="‚òÄÔ∏è"
                className="theme-switch"
                aria-label="Cambiar tema"
            />



            {/* Bot√≥n para GitHub */}
            <Button
                size="middle"
                icon={<GithubOutlined className="brand-icon brand-icon--github" />}
                href="https://github.com/Excintium"
                target="_blank"
                className="site-action site-action--github"
                aria-label="GitHub"
            >
                GitHub
            </Button>

            {/* Bot√≥n para LinkedIn */}
            <Button
                size="middle"
                icon={<LinkedinFilled className="brand-icon brand-icon--linkedin" />}
                href="https://www.linkedin.com/in/nicol%C3%A1s-fonseca"
                target="_blank"
                className="site-action site-action--linkedin"
                aria-label="LinkedIn"
            >
                LinkedIn
            </Button>
        </Space>
    );

    // ---- RENDERIZADO ----
    return (
        // Uso la etiqueta 'header' de HTML5.
        // Le aplico los estilos base que vienen de 'app/styles.tsx'.
        <header className="site-header" style={{ ...styles.headerBar, margin: 0, padding: 0 }}>
            {/* Este 'div' es el contenedor centrado (max-width) */}
            <div
                className="container"
                style={{
                    ...styles.container,
                    height: 64, // Altura fija para el header
                    display: "flex",
                    alignItems: "center",
                    justifyContent: "space-between", // Separa logo de acciones
                    gap: 12,
                }}
            >
                {/* IZQUIERDA: Mi logo "Excintium" */}
                <a
                    href="#inicio"
                    style={{
                        display: "flex",
                        alignItems: "center",
                        gap: 10,
                        textDecoration: "none",
                        color: "inherit",
                        position: "relative",
                    }}
                    aria-label="Ir al inicio"
                >
                    <span className="brand-logo" aria-label="GengarChikito">GengarChikito</span>
                </a>

                {/* CENTRO: Navegaci√≥n (Solo en escritorio) */}
                {/* Renderizado condicional: Si NO es m√≥vil,
                  muestro la variable 'navLinks' en el centro.
                */}
                {!isMobile && (
                    <nav style={{ flex: 1, display: "flex", justifyContent: "center" }}>
                        {navLinks}
                    </nav>
                )}

                {/* DERECHA: Botones de Acci√≥n (Escritorio) o Bot√≥n de Men√∫ (M√≥vil) */}
                {!isMobile ? (
                    // Si NO es m√≥vil (escritorio), muestro los botones de acci√≥n.
                    <div style={{ display: "flex", alignItems: "center", gap: 12 }}>
                        {actionButtons}
                    </div>
                ) : (
                    // Si S√ç es m√≥vil, muestro el bot√≥n de hamburguesa.
                    <Button
                        aria-label="Abrir men√∫"
                        icon={<MenuOutlined />}
                        onClick={() => setOpen(true)} // Al hacer clic, abro el Drawer
                        className="site-action"
                    />
                )}
            </div>

            {/* Panel/Men√∫ lateral que se abre en m√≥vil */}
            <Drawer
                open={open} // Abierto o cerrado seg√∫n mi estado 'open'
                onClose={() => setOpen(false)} // Funci√≥n para cerrarlo (ej. al hacer clic fuera)
                placement="right"
                width={320}
                styles={{
                    // Aqu√≠ uso las variables CSS que defin√≠ en ThemedWrapper
                    // para que el Drawer tambi√©n cambie con el tema.
                    body: { padding: 16, background: "var(--surface-bg)" },
                    header: { borderBottom: "1px solid var(--border-color)" },
                }}
            >
                {/* Contenido del Drawer */}
                <Space direction="vertical" size={16} style={{ width: "100%" }}>

                    {/* El Switch de Tema, pero dentro del men√∫ m√≥vil */}
                    <div
                        style={{
                            display: "flex",
                            justifyContent: "space-between",
                            alignItems: "center",
                        }}
                    >
                        <Typography.Text style={{ fontWeight: 600 }}>Tema</Typography.Text>
                        <Switch
                            checked={dark}
                            onChange={onToggleDark}
                            checkedChildren="üåô"
                            unCheckedChildren="‚òÄÔ∏è"
                            className="theme-switch"
                            aria-label="Cambiar tema"
                        />
                    </div>

                    {/* La lista de navegaci√≥n, pero vertical */}
                    <div style={{ display: "flex", flexDirection: "column", gap: 12 }}>
                        {/* Les a√±ado un 'onClick' a los enlaces del Drawer.
                          Cuando hago clic en uno, navega a la secci√≥n Y
                          tambi√©n llama a 'setOpen(false)' para cerrar el men√∫.
                        */}
                        <a className="site-nav-link" href="#inicio" onClick={() => setOpen(false)}>Inicio</a>
                        <a className="site-nav-link" href="#proyectos" onClick={() => setOpen(false)}>Proyectos</a>
                        <a className="site-nav-link" href="#experiencia" onClick={() => setOpen(false)}>Experiencia</a>
                        <a className="site-nav-link" href="#educacion" onClick={() => setOpen(false)}>Educaci√≥n</a>
                        <a className="site-nav-link" href="#contacto" onClick={() => setOpen(false)}>Contacto</a>
                    </div>

                    <div style={{ height: 8 }} />

                    {/* Aqu√≠ reutilizo la variable 'actionButtons'
                       para mostrar los botones de CV, GitHub y LinkedIn
                       tambi√©n dentro del men√∫ m√≥vil.
                    */}
                    {actionButtons}
                </Space>
            </Drawer>
        </header>
    );
}