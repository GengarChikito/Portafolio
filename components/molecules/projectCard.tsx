import { Card, Space, Tag } from "antd";

/**
 * Defino las 'props' que mi ProjectCard necesita.
 * Esto me obliga a pasarle toda la información
 * correcta desde el objeto 'profile'.
 */
export interface ProjectCardProps {
    /** El título que se mostrará en la tarjeta. */
    title: string;
    /** La descripción corta (summary) del proyecto. */
    summary: string;
    /** Un array de strings (ej. ["React", "TypeScript"]) para las 'Tags'. */
    tags: string[];
    /** La URL (http o data-uri) de la imagen de portada. */
    image: string;
    /** El enlace externo (ej. a GitHub o Vercel) al que irá la tarjeta. */
    link: string;
}

/**
 * Este es mi componente 'ProjectCard' (molécula).
 * Lo diseño para que sea una tarjeta clickeable que muestra
 * la vista previa de uno de mis proyectos.
 */
export function ProjectCard({ title, summary, tags, image, link }: ProjectCardProps) {
    return (
        // Envuelvo toda la 'Card' en un enlace '<a>' para que sea
        // completamente clickeable y lleve al 'link' del proyecto.
        <a href={link} target="_blank" rel="noreferrer">
            <Card
                hoverable // Activo el efecto de sombra de Ant Design al pasar el mouse.
                className="project-card" // Una clase CSS para estilos personalizados (como el tamaño de la img)

                // 'cover' es una prop de Ant Design para poner una imagen en la cabecera.
                cover={
                    <img
                        src={image}
                        alt={title} // El 'alt' text es crucial para accesibilidad.
                        /* El tamaño de esta imagen (200px) lo defino en app.css
                           usando la clase .project-card img {} */
                    />
                }
                // Aplico estilos para bordes redondeados y un borde que usa
                // mi variable CSS (--border-color) para que cambie con el tema.
                style={{
                    borderRadius: 16,
                    border: "1px solid var(--border-color, #E5E7EB)",
                }}
            >
                {/* 'Card. Meta' es un sub-componente de Ant Design
                    perfecto para mostrar un título y una descripción. */}
                <Card.Meta title={title} description={summary} />

                {/* Uso un componente 'Space' de Ant Design para apilar
                    mis 'Tags' y que se ajusten ('wrap') si no caben. */}
                <Space wrap size={[6, 6]} style={{ marginTop: 12 }}>
                    {/* Hago un .map() sobre el array 'tags' que recibí en las props
                        y renderizo un componente <Tag> por cada uno. */}
                    {tags.map((t) => (
                        <Tag key={t}>{t}</Tag>
                    ))}
                </Space>
            </Card>
        </a>
    );
}