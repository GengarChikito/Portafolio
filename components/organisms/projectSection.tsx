// Importo mis estilos, mi "base de datos" de perfil, y mis átomos/moléculas
import { styles } from "app/styles";
import { profile } from "components/atoms/profile/profile";
import { SectionTitle } from "components/atoms/sectionTitle";
// Importo la molécula 'ProjectCard' que definí antes
import { ProjectCard } from "../molecules/projectCard"; // (Asegúrate que la ruta '../molecules' sea correcta)

/**
 * Defino las 'props' para esta sección.
 * No necesito ninguna, ya que esta sección (siendo un "organismo")
 * sabe de dónde sacar sus datos: directamente del 'profile' importado.
 */
export interface ProjectsSectionProps {}

/**
 * Este es mi componente 'ProjectsSection' (un "organismo").
 * Su trabajo es simple:
 * 1. Renderizar el título de la sección.
 * 2. Mapear (hacer un bucle) sobre la lista de proyectos en mi 'profile'.
 * 3. Renderizar un componente 'ProjectCard' (molécula) por cada proyecto.
 */
export function ProjectsSection(_: ProjectsSectionProps) {
    return (
        // La <section> principal con el 'id' para la navegación
        <section id="proyectos" style={styles.section}>
            <div className="container" style={styles.container}>

                {/* 1. Renderizo mi átomo de título */}
                <SectionTitle text="Proyectos" />

                {/* 2. Este 'div' usa una clase CSS ('project-grid')
                       que definí en 'app.css' para crear la cuadrícula responsive.
                 */}
                <div className="project-grid">

                    {/* 3. Aquí hago el .map() sobre mi array 'profile.projects' */}
                    {profile.projects.map((p) => (
                        // Por cada 'p' (proyecto) en el array, renderizo mi molécula 'ProjectCard'
                        <ProjectCard
                            key={p.title} // React necesita una 'key' única para cada ítem
                            // Le paso todas las propiedades del objeto 'p'
                            // como 'props' al componente 'ProjectCard'.
                            title={p.title}
                            summary={p.summary}
                            tags={p.tags}
                            image={p.image}
                            link={p.link}
                        />
                    ))}
                </div>
            </div>
        </section>
    );
}