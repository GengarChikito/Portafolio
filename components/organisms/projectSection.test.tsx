import React from "react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";


// Mock de estilos
vi.mock("app/styles", () => ({
    styles: {
        section: { border: "1px solid purple" },
        container: { maxWidth: 1200 },
    },
}));

// Mock del perfil con datos de proyectos
vi.mock("components/atoms/profile/profile", () => {
    const mockProjects = [
        {
            title: "Proyecto Alpha",
            summary: "Descripción Alpha",
            tags: ["React", "Node"],
            image: "/alpha.jpg",
            link: "https://alpha.com",
        },
        {
            title: "Proyecto Beta",
            summary: "Descripción Beta",
            tags: ["Vue", "Firebase"],
            image: "/beta.jpg",
            link: "https://beta.com",
        },
    ];
    return {
        profile: {
            projects: mockProjects,
        },
    };
});

// Mock de SectionTitle
vi.mock("components/atoms/sectionTitle", () => ({
    SectionTitle: ({ text }: { text: string }) => (
        <h2 data-testid="section-title">{text}</h2>
    ),
}));

// Mock *general* de ProjectCard
const projectCardSpy = vi.fn();
// Importamos ProjectCard aquí para poder usar vi.mocked después
import { ProjectCard as OriginalProjectCard } from "../molecules/projectCard"; // Renombramos para evitar conflicto
vi.mock("../molecules/projectCard", () => ({
    ProjectCard: vi.fn((props: any) => { // Usamos vi.fn() para poder espiar y reimplementar
        projectCardSpy(props);
        return <div data-testid="project-card">{props.title}</div>;
    }),
}));

// ====== SUT ======
import { ProjectsSection } from "./projectSection";


// Obtenemos una referencia al mock de ProjectCard
const MockedProjectCard = vi.mocked(OriginalProjectCard); // Usamos el nombre original importado

// Limpiar mocks antes de cada prueba
beforeEach(() => {
    vi.clearAllMocks();
    projectCardSpy.mockClear();
    // Aseguramos que el mock use la implementación por defecto antes de cada test
    MockedProjectCard.mockImplementation((props: any) => {
        projectCardSpy(props);
        return <div data-testid="project-card">{props.title}</div>;
    });
});

// ====== Tests ======
describe("projectsSection", () => {
    it("renderiza el título de la sección", () => {
        render(<ProjectsSection />);
        const title = screen.getByTestId("section-title");
        expect(title).toBeInTheDocument();
        expect(title).toHaveTextContent("Proyectos");
    });

    it("renderiza una ProjectCard para cada proyecto del perfil", async () => {
        const { profile } = vi.mocked(await import("components/atoms/profile/profile"));
        render(<ProjectsSection />);
        const cards = screen.getAllByTestId("project-card");
        expect(cards).toHaveLength(profile.projects.length);
        expect(screen.getByText("Proyecto Alpha")).toBeInTheDocument();
        expect(screen.getByText("Proyecto Beta")).toBeInTheDocument();
    });

    it("pasa las props correctas a cada ProjectCard", async () => {
        const { profile } = vi.mocked(await import("components/atoms/profile/profile"));
        render(<ProjectsSection />);

        expect(projectCardSpy).toHaveBeenCalledWith(
            expect.objectContaining({
                title: profile.projects[0].title,
                summary: profile.projects[0].summary,
                tags: profile.projects[0].tags,
                image: profile.projects[0].image,
                link: profile.projects[0].link,
            })
        );

        expect(projectCardSpy).toHaveBeenCalledWith(
            expect.objectContaining({
                title: profile.projects[1].title,
                summary: profile.projects[1].summary,
                tags: profile.projects[1].tags,
                image: profile.projects[1].image,
                link: profile.projects[1].link,
            })
        );
    });

    // ==== INICIO DE LA CORRECCIÓN ====
    // Ya no usamos vi.mock aquí dentro
    it("snapshot del componente", () => {
        // Cambiamos la implementación *solo para esta prueba*
        MockedProjectCard.mockImplementationOnce((props: any) => {
            projectCardSpy(props); // Mantenemos el espía si es necesario
            // Renderizamos con data-key para el snapshot
            return <div data-testid="project-card" data-key={props.key}>{props.title}</div>;
        });
        // La segunda card también necesita esta implementación especial
        MockedProjectCard.mockImplementationOnce((props: any) => {
            projectCardSpy(props);
            return <div data-testid="project-card" data-key={props.key}>{props.title}</div>;
        });


        const { container } = render(<ProjectsSection />);
        expect(container).toMatchSnapshot();

    });
});