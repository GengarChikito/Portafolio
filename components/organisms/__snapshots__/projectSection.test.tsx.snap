// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`projectsSection > snapshot del componente 1`] = `
<div>
  <section
    id="proyectos"
    style="border: 1px solid purple;"
  >
    <div
      class="container"
      style="max-width: 1200px;"
    >
      <h2
        data-testid="section-title"
      >
        Proyectos
      </h2>
      <div
        class="project-grid"
      >
        <div
          data-testid="project-card"
        >
          Proyecto Alpha
        </div>
        <div
          data-testid="project-card"
        >
          Proyecto Beta
        </div>
      </div>
    </div>
  </section>
</div>
`;
